src/celeritas/optical/CerenkovPreGenerator.hh

```diff
@@ -0,0 +1,139 @@
+
+#include "CerenkovData.hh"
+#include "CerenkovDndxCalculator.hh"
+#include "OpticalDistributionData.hh"
+#include "OpticalPropertyData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the number of Cerenkov photons to be generated by
+ * \c CerenkovGenerator and populate \c OpticalDistributionData values using
+ * Param and State data.
+ * \code
+    CerenkovPreGenerator pre_cerenkov(particle_view,
+                                      properties,
+                                      cerenkov_data,
+                                      material_id,
+                                      optical_step_collector.get(tid));
+
+    auto optical_dist_data = pre_cerenkov(rng);
+    if (optical_dist_data)
+    {
+        CerenkovGenerator cerenkov_gen(... , optical_dist_data, ...);
+        cerenkov_gen(rng);
+    }
+ * \endcode
+ */
+class CerenkovPreGenerator
+{
+  public:
+    // Construct with particle and material data
+    inline CELER_FUNCTION
+    CerenkovPreGenerator(ParticleTrackView const& particle_view,
+                         NativeCRef<OpticalPropertyData> const& properties,
+                         NativeCRef<CerenkovData> const& shared,
+                         OpticalMaterialId mat_id,
+                         OpticalStepCollectorData const& optical_step);
+
+    // Return a populated distribution data for the Cerenkov Generator
+    template<class Generator>
+    inline CELER_FUNCTION OpticalDistributionData operator()(Generator& rng);
+
+  private:
+    units::ElementaryCharge charge_;
+    OpticalMaterialId mat_id_;
+    OpticalStepCollectorData step_collector_data_;
+    real_type num_photons_per_len_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from particle, material, and optical step collection data.
+ */
+CELER_FUNCTION CerenkovPreGenerator::CerenkovPreGenerator(
+    ParticleTrackView const& particle_view,
+    NativeCRef<OpticalPropertyData> const& properties,
+    NativeCRef<CerenkovData> const& shared,
+    OpticalMaterialId mat_id,
+    OpticalStepCollectorData const& optical_step)
+    : charge_(particle_view.charge())
+    , mat_id_(mat_id)
+    , step_collector_data_(optical_step)
+{
+    CELER_EXPECT(charge_.value() != 0);
+    CELER_EXPECT(mat_id_);
+    CELER_EXPECT(step_collector_data_);
+
+    auto const& pre = step_collector_data_.points[StepPoint::pre];
+    auto const& post = step_collector_data_.points[StepPoint::post];
+    units::LightSpeed beta(real_type{0.5}
+                           * (pre.speed.value() + post.speed.value()));
+
+    CerenkovDndxCalculator dndx_calculator(
+        properties, shared, mat_id_, charge_);
+    num_photons_per_len_ = dndx_calculator(beta);
```

Comment:
Try to phrase functor instances as verbs, just like you would name a function: an instance of `FooCalculator` should be `calculate_foo`.